### ISSUE TITLE
Define quote escaping semantics.

### ISSUE THREAD

**Issue by jordansissel** [elastic/logstash#1645]
This is requested numerous places, but I was reminded from https://github.com/elasticsearch/logstash/pull/895

Escaping quotes and control characters are impossible to do today in Logstash config. "\n" is literally backslash and lowercase n. "\"" is literally backslash and doublequote.

Logstash supports single and double-quoted text values, and we should support escapes in both the same exact way:
## Proposal

Add a new setting, `config.strings.support_escapes` that, when set, will enable the following behaviors on quoted strings in the Logstash config:

- quote-backslash-quote-quote should be text with a single quote as value
- doublequote-backslash-doublequote-doublequote should be text with a single double-quote as value
- Because it is expected by users, "\t" and '\t' should be a single tab character as value (ASCII 09)
- Because it is expected by users, "\n" and '\n' should be a single line feed character as value (ASCII 10)
- Because of \t and \n, we should also support "\r" and '\r' as a single carriage return (ASCII 13)
- "\\\\" and '\\\\' should be text with a single backslash (written as two backslashes in the config)
- Unicode text is fully supported and does not need escaping

## Workarounds today:
- double-quote: Use single-quoted strings: `setting => 'I said, "Hello!"'`
- single-quote: Use double-quoted strings: `setting => "I said, 'Hello!'"`
- backslash: Use a single backslash: `setting => "C:\some\path"`
- tab: use an actual tab character. In some terminal-based editors, you can insert an actual tab character by pressing Control+V then pressing Tab afterwards.
- newline: use an actual newline character: 

```
setting => "this
is
multiple
lines"
```

**Comment by splashx** [comment]
is this issue related to [this issue](https://github.com/logstash-plugins/logstash-filter-mutate/issues/40)?

**Comment by jordansissel** [comment]
@splashx yep :)

**Comment by splashx** [comment]
@jordansissel oh no :laughing:

**Comment by acchen97** [comment]
+1 as I just hit this today too

**Comment by bcubk** [comment]
Hi @jordansissel, are there any progress about this issue? Is there a timetable when it's fixed?

**Comment by nellicus** [comment]
+1 hit today (same as https://github.com/elastic/logstash/issues/3239 6 months ago)

**Comment by jordansissel** [comment]
This change is likely to be resolved eventually, but not right now. There's a lot of work going on for the new centralized configuration feature that's coming soon and will impact how we store and represent the configuration of a logstash agent. I'd rather fix this problem at _that_ time than try to fix it now and break everyone who is using the current not-so-good behavior.

**Comment by bcubk** [comment]
Thank you for this information :+1:

**Comment by jhftrifork** [comment]
I'm not sure what to make of this. Could someone confirm that this is the current behavior?

Upon encountering a quote character, the Logstash config parser interprets this as a string literal, which is terminated by the next quote character of the same type as the opening quote. This string literal then represents the string of characters between the start and end quote characters in the surface syntax.

This means that a character sequence `seq` represents a string `str` if either

* `str` contains no single-quotes, and `seq = "'" + str + "'"`, or
* `str` contains no double-quotes, and `seq = '"' + str + '"'`.

If `str` contains single-quotes and double-quotes, it is unrepresentable as a string literal in Logstash syntax. (Deal with it.)

**Comment by sts** [comment]
@jordansissel got any updates on when we could reckon on a solution to this issue?

**Comment by jordansissel** [comment]
@sts We can probably have it ready for the next major release of Logstash (early next year) since that'll be the next time we can really introduce breaking changes.

**Comment by jordansissel** [comment]
@suyograo @acchen97 - I think we can ship this improvement with the next major given all the new work going into configuration features.

**Comment by jordansissel** [comment]
@sts As a short-term alternative, I'd be willing to add a flag to enables escaping support in the config processor. This would maintain backwards compat _and_ make it available in the Logstash 2.x series.

**Comment by sts** [comment]
@jordansissel  that would be awesome!

**Comment by wenerme** [comment]
How can I work around this ? by using ruby code directly ?

**Comment by loms0815** [comment]
Hi @suyograo, 
can you tell the name of the hidden flag, which change the quote escaping.
Thank you in advance!

**Comment by mmatson** [comment]
+1 Would be lovely to have this behave as expected by commonly accepted standards. Makes for painful on-boarding of more users in my organization as developers struggle to figure out that string manipulation in the log file parsing is challenging.

The idea of a flag for switching behavior works for me.

**Comment by jonathanpmartins** [comment]
👍

**Comment by antonin42** [comment]
+1 for this issue, following the configuration page, setting delimiter to it's theoritical default value [(according to official documentation)](https://www.elastic.co/guide/en/logstash/current/plugins-inputs-file.html#plugins-inputs-file-delimiter) does not work.


`delimiter => "\n"`

**Comment by andrewvc** [comment]
There's a hack around this using the ruby filter and ASCII ordinals https://gist.github.com/andrewvc/1b6e5c72cf124ed39d311f9714dac271

**Comment by pgmillon** [comment]
Any update on this ? This issue has been opened for more than 2 years now.

**Comment by jordansissel** [comment]
No update at this time.

**Comment by jordansissel** [comment]
I've updated the issue's description to include some possible workarounds for specific characters. See the section under "Workarounds today:"

**Comment by jordansissel** [comment]
@suyograo @acchen97 I'm open to adding this as an off-by-default feature as soon as maybe Logstash 5.2 or 5.3 timeframe.

**Comment by acchen97** [comment]
@jordansissel +1 let's try to target 5.2 for this. Can you please own this feature?

**Comment by frittentheke** [comment]
Glad this issue is somewhat on your radar. The existing workarounds are helpful, but one thing is apparently still impossible to do with the current configuration parser:  Using mutate+gsub to replace something matched with a double-quote character . i.e.

`
mutate {
  gsub => [
   'message', '\"','"'
  ]
}   
`

**Comment by jakommo** [comment]
@jordansissel @acchen97 any new on this?

**Comment by jordansissel** [comment]
No news. There are many workarounds that can still be used today. I want to
add this feature, but we are overloaded with tasks as it is and it may take
some time before this gets worked on.

**Comment by binary132** [comment]
This makes it impossible to insert e.g. control characters.

**Comment by jordansissel** [comment]
I have a branch where I am working on this as time permits. No ETA. https://github.com/elastic/logstash/compare/feature/config-string-escapes

**Comment by binary132** [comment]
I was able to work around this by manually inserting the character literal. 😄

**Comment by jordansissel** [comment]
PR is open for this: https://github.com/elastic/logstash/pull/7442

**Comment by suyograo** [comment]
Fixed in #7442

**Comment by Mekk** [comment]
Hmm. I configure fresh ELK system and it took me few days to find why my text contain \n (lteral \ then literal n). 

Well, I wrote 
```
mutate {  join => "somefield" => "\n" }
```
